### PR TITLE
Moved particle index into CellStructure

### PR DESCRIPTION
**Pull request by fweik**
Description of changes:
- Added unified interface to add and remove particles to `CellStructure`
- Local particle index is now maintained by the cell system.

### CODE REVIEW

File: src/core/CMakeLists.txt [hunk 1]
```diff
@@ -42,7 +41,7 @@ set(EspressoCore_SRC
     thermostat.cpp
     tuning.cpp
     virtual_sites.cpp
-    exclusions.cpp)
+    exclusions.cpp CellStructure.cpp CellStructure.hpp)
```
Comment (jngrad) [inline src/core/CMakeLists.txt]: diffs are more readable with one element per line, in my opinion:
```suggestion
    exclusions.cpp
    CellStructure.cpp
    CellStructure.hpp
)
```

File: src/core/CellStructure.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
+#include "CellStructure.hpp"
```
Comment (jngrad) [inline src/core/CellStructure.cpp:22]: missing license header, needs to be copy-pasted from `src/core/cells.cpp`

File: src/core/CellStructure.hpp [hunk 3]
```diff
@@ -0,0 +1,245 @@
+#ifndef ESPRESSO_CELLSTRUCTURE_HPP
```
Comment (jngrad) [inline src/core/CellStructure.hpp:22]: missing license header, needs to be copy-pasted from `src/core/cells.hpp`

File: src/core/io/mpiio/mpiio.cpp [hunk 4]
```diff
@@ -48,16 +48,13 @@
...
+#include "mpiio.hpp"
 #include "Particle.hpp"
 #include "bonded_interactions/bonded_interaction_data.hpp"
 #include "cells.hpp"
 #include "errorhandling.hpp"
 #include "event.hpp"
-#include "mpiio.hpp"
-#include "particle_data.hpp"
-#include "particle_index.hpp"
+//#include "particle_data.hpp"
```
Comment (jngrad) [inline src/core/io/mpiio/mpiio.cpp]: can this commented out include statement be removed?

**Review by jngrad — CHANGES_REQUESTED** [review]
There are a few parts of the code I'm not too familiar with (e.g. ghosts), but overall this refactor makes the cell structure framework a lot less confusing. I only have minor changes to propose for now.

File: src/core/CMakeLists.txt (hunk 1 above)
Comment (fweik) [reply src/core/CMakeLists.txt]: There are no rules for cmake formatting ^^

**Comment by fweik** [comment]
If you have questions about he code don't hesitate to ask. Might also be a pointer to where the code should be improved.

File: src/core/CellStructure.hpp [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+  std::vector<Particle *> m_particle_index;
+
+public:
+  /**
+   * @brief Update local particle index.
+   *
+   * Update the entry for a particle in the local particle
+   * index.
+   *
+   * @param id Entry tto update.
```
Comment (jngrad) [inline src/core/CellStructure.hpp]:
```suggestion
   * @param id Entry to update.
```

File: src/core/CellStructure.hpp [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    if (id >= m_particle_index.size())
+      m_particle_index.resize(id + 1);
+
+    m_particle_index[id] = p;
+  }
+
+  /**
+   * @brief Update local particle index.
+   *
+   * @param pl List of particle whose index entries should be updated.
```
Comment (jngrad) [inline src/core/CellStructure.hpp]:
```suggestion
   * @param pl List of particles whose index entries should be updated.
```

File: src/core/CellStructure.hpp [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+  void update_particle_index(ParticleList &pl) {
+    for (auto &p : pl) {
+      update_particle_index(p.identity(), std::addressof(p));
+    }
+  }
+
+  /**
+   * @brief Update local particle index.
+   *
+   * @param pl List of particle whose index entries should be updated.
```
Comment (jngrad) [inline src/core/CellStructure.hpp]:
```suggestion
   * @param pl List of particles whose index entries should be updated.
```

File: src/core/ParticleList.hpp [hunk 8]
```diff
@@ -87,6 +93,46 @@ struct ParticleList {
...
+   *
+   * @return Last particle in the list.
+   */
+  Particle &&extract_back() { return std::move(part[--n]); }
+
+  /**
+   * @brief Move out the i-th particle in the list.
+   *
+   *  This can change the order of particles
+   *  at from position i onward (including i).
```
Comment (jngrad) [inline src/core/ParticleList.hpp]:
```suggestion
   *  from position i onward (including i).
```

File: src/core/cells.cpp [hunk 9]
```diff
@@ -338,35 +332,52 @@ ParticleList sort_and_fold_parts(const CellStructure &cs, CellPList cells) {
...
     break;
   case CELL_STRUCTURE_DOMDEC:
-    dd_exchange_and_sort_particles(global_flag, &displaced_parts, node_grid);
+    dd_exchange_and_sort_particles(global_flag, &displaced_parts, node_grid,
+                                   modified_cells);
     break;
   }
 
+  boost::sort(modified_cells);
+  for (auto cell : modified_cells | boost::adaptors::uniqued) {
```
Comment (jngrad) [inline src/core/cells.cpp:364]: Bash syntax in C++ code. What a time to be alive :smile:

File: src/core/virtual_sites/lb_inertialess_tracers.cpp [hunk 10]
```diff
@@ -116,12 +116,10 @@ void IBM_UpdateParticlePositions(ParticleRange particles) {
 #endif
         p.r.p[2] = p.r.p[2] + p.m.v[2] * time_step;
 
-      // Check if the particle might have crossed a box border (criterion see
-      // e-mail Axel 28.8.2014)
```
Comment (jngrad) [inline src/core/virtual_sites/lb_inertialess_tracers.cpp:120]: that's so meta

**Review by jngrad — COMMENTED** [review]
Besides the typos in old comments, this looks good.

File: src/core/cells.cpp (hunk 9 above)
Comment (fweik) [reply src/core/cells.cpp:364]: With C++20 all code will look like this :-)

Comment (jngrad) [reply src/core/cells.cpp:364]: I looked up the Boost docs on Ranges and found how to pipe data to grep-expressions and sed-expressions, Bash style:
```c++
#include <boost/range/adaptor/transformed.hpp>
#include <boost/range/adaptor/filtered.hpp>
#include <boost/range/algorithm/copy.hpp>
#include <iterator>
#include <iostream>
#include <vector>
#include <string>
#include <regex>

struct grep_expression {
  bool operator()(std::string const &s) const {
      return std::regex_match(s, std::regex("[A-Z]{2}.*"));
  }
};

struct sed_expression {
  std::string operator()(std::string const &s) const {
    return std::regex_replace(s, std::regex("ESPResSo"), "CAPPucciNo");
  }
};

int main() {
    using namespace boost::adaptors;
    std::vector<std::string> input = {"ESPResSo", "Python", "LB"};
    boost::copy(
        input | filtered(grep_expression()) | transformed(sed_expression()),
        std::ostream_iterator<std::string>(std::cout, "\n"));
    return 0;
}
```

**Review by jngrad — APPROVED** [review]

**Pull request merged**